src/qibo/models/variational.py

```diff
@@ -96,6 +96,135 @@ def _loss(params, circuit, hamiltonian):
         return result, parameters, extra
 
 
+class AAVQE(object):
+    """This class implements the Adiabatically Assisted Variational Quantum Eigensolver
+    algorithm. See https://arxiv.org/abs/1806.02287.
+
+    Args:
+        circuit (:class:`qibo.abstractions.circuit.AbstractCircuit`): variaional ansatz.
+        easy_hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): initial Hamiltonian object.
+        problem_hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): problem Hamiltonian object.
+        s (callable): scheduling function of time that defines the adiabatic
+            evolution. It must verify boundary contitions: s(0) = 0 and s(1) = 1.
+        nsteps (float): number of steps of the adiabatic evolution.
+        t_max (float): total time evolution.
+        bounds_tolerance (float): tolerance for checking s(0) = 0 and s(T) = 1.
+        time_tolerance (float): tolerance for checking if time is greater than t_max.
+
+    Example:
+    ::
+        import numpy as np
+        from qibo import gates, models, hamiltonians
+        # create circuit ansatz for two qubits
+        circuit = models.Circuit(2)
+        circuit.add(gates.RY(0, theta=0))
+        circuit.add(gates.RY(1, theta=0))
+        # define the easy and the problem Hamiltonians.
+        easy_hamiltonian=hamiltonians.X(2)
+        problem_hamiltonian=hamiltonians.XXZ(2)
+        # define a scheduling function with only one parameter 
+        # and boundary conditions s(0) = 0, s(1) = 1
+        s = lambda t: t
+        # create AAVQE model 
+        aavqe = models.AAVQE(circuit, easy_hamiltonian, problem_hamiltonian, 
+                             s, nsteps=10, t_max=1)
+        # optimize using random initial variational parameters
+        np.random.seed(0)
+        initial_parameters = np.random.uniform(0, 2*np.pi, 2)
+        ground_energy, params = aavqe.minimize(initial_parameters)
+    """
+
+    def __init__(self, circuit, easy_hamiltonian, problem_hamiltonian, s, nsteps=10, t_max = 1, bounds_tolerance = 1e-7, time_tolerance = 1e-7):
+
+        if nsteps <= 0: # pragma: no cover
+            raise_error(ValueError, "Number of steps nsteps should be positive but is {}."
+                                    "".format(nsteps))
+        if t_max <= 0: # pragma: no cover
+            raise_error(ValueError, "Maximum time t_max should be positive but is {}."
+                                    "".format(t_max))
+        if easy_hamiltonian.nqubits != problem_hamiltonian.nqubits: # pragma: no cover
+            raise_error(ValueError, "The easy Hamiltonian has {} qubits while problem Hamiltonian has {}."
+                                    "".format(easy_hamiltonian.nqubits, problem_hamiltonian.nqubits))
```

Comment:
We may also implement some tests for the errors and remove ``#  pragma: no cover`` from these if statements, but maybe it's excessive. The same holds for all the other checks.